### ISSUE TITLE
Quick Choice FSC: Enhancement of Visual Cards with custom images

### ISSUE THREAD

**Issue by clifford-fra** [alexed1/LightningFlowComponents#252]
Currently, the choice Icons (String collection) input only allows the usage of SF Icons in the form of e.g. utility:activity.

It would be nice, if we could (also) use images, which are uploaded as static resources.

**Comment by ericrsmith35** [comment]
Another future enhancement.  Any takers?

**Comment by clifford-fra** [comment]
Hi Eric,

I was inspired by your blog article for the "quick smart app switcher":
https://ericsplayground.wordpress.com/2020/02/22/add-a-new-quick-smart-app-switcher/

I looked at the code, reverse engineered a bit and was finally able to implement custom images as visual card icons.

Here's how I have it done:

1. Add a zip file as a static resource with the name `Quickchoice_Images`. Our custom images will be placed inside there. If you don't want to use custom images and stay with the SF icons, then just leave the zip file empty.

2. At first, lets work on the **quickChoiceFSC.js**. We need to add an import at the beginning:

`import Quickchoice_Images from '@salesforce/resourceUrl/Quickchoice_Images';`

3. Then we need to add the correct path to the images in the **quickChoiceFSC.js**, while also preserving the ability to leave SF icons unchanged. So we replace this code snippet:

```
this.choiceLabels.forEach(label => {
	items.push({ name: label, description: this.choiceValues[index], icon: this.choiceIcons[index] });
	...
});
```

with this one:

```
this.choiceLabels.forEach(label => {
	//Add the correct path to custom images
	if (this.choiceIcons[index].includes(':')) {
		items.push({ name: label, description: this.choiceValues[index], icon: this.choiceIcons[index] });
	} else {
		items.push({ name: label, description: this.choiceValues[index], icon: Quickchoice_Images + '/' + this.choiceIcons[index] });
	}
	...
});
```
This just means, that in case a SF Icon in the form of `custom:custom20` is passed to the component by the user, the script will leave it as it is. But if an image is passed e.g. `bank.png`, then it will be `Quickchoice_Images/bank.png`.

4. Let's look at the **quickChoiceFSC.html** file. The magic happens here. In the section  `<!-- Display Visual Card Pickers with Icons-->` there's this code for inserting the icon:
```
<span class="slds-icon_container">
	<lightning-icon icon-name={item.icon} size={iconSize}>
	</lightning-icon>
</span>
```

we will replace it with one code line:
`<lightning-avatar size={iconSize} src={item.icon} fallback-icon-name={item.icon}></lightning-avatar>`

Finished and everything is done! The lightning avatar uses the fallback-icon in order to still render the icon, if the src cannot be found. If an image is passed, it will just render the image :-) Using `<lightning-avatar>` allows us to mix custom images and sf icons.

![example](https://user-images.githubusercontent.com/55443386/75336659-d1b1a480-588b-11ea-95b5-3d44ecbd0ba4.png)

Pros of this solution:
- The component interface remains the same.
- The choice Icons input variable can consume something like `standard:entity_milestone, bank.png, custom:custom20`.
- Users of this component do not need to change anything. (except if icon size is xx-small)

Cons:
- We need to add an empty zip file with the name `Quickchoice_Images` as a static resource to the package.
- The `<lightning-avatar>` tag does not support `xx-small` as icon size. Only `x-small, small, medium and large` are supported. But honestly, I don't see a need for `xx-small` in this component.

I would like to here your opinion about this solution.

Best regards
Clifford

**Comment by ericrsmith35** [comment]
Nice work Clifford.  I wonder if we should suggest using and code for the same Static Resource image file for both components?

**Comment by clifford-fra** [comment]
I would suggest to leave them separate in order to avoid any conflicts. I will do some testing later on a fresh playground. If everything works as expected, we can consider a new version of the Quickchoice component.

**Comment by ericrsmith35** [comment]
I'm also thinking it wouldn't hurt to add the ability to use an icon in the
App Switcher.

**Comment by clifford-fra** [comment]
Hi Eric, I've finished my tests. Everything works and looks as expected. I have also tested successfully with lightning out on an external page. A few things to consider:

- If icon size `xx-small` is used, the size will be set to medium by default without any user interaction.

- `xx-small` icon size is not supported anymore. Encouraged is the use of `x-small` (Only 2px difference in width and height).

- If an invalid image file name is passed to the component, the component will render whitespace. This is OK and no error is thrown.

I think we are good to go for a new update. Custom images with a large size look great!

Best regards, Clifford
